### ISSUE TITLE
paste performance issues

### ISSUE THREAD

**Issue by brokedarius** [xonsh/xonsh#5292]
## xonfig

<details>

```
import datetime
from prompt_toolkit import prompt
from prompt_toolkit.history import InMemoryHistory
from prompt_toolkit.key_binding import KeyBindings
from prompt_toolkit.keys import Keys

bindings = KeyBindings()

@bindings.add(Keys.Up)
def history_search_backward(event):
    event.current_buffer.history_backward()

@bindings.add(Keys.Down)
def history_search_forward(event):
    event.current_buffer.history_forward()

$PROMPT = '{INTENSE_GREEN}[{YELLOW}{user}{RESET}@{BLUE}{hostname}{RESET}:{cwd}─{INTENSE_YELLOW}[' + str(datetime.date.today()) + ']─{INTENSE_GREEN}[{localtime}]{INTENSE_RED}──>{RESET} '
$XONSH_HISTORY_BACKEND = 'sqlite'
$FORCE_POSIX_PATHS = True
$XONSH_SHOW_TRACEBACK = True
$INTENSIFY_COLORS_ON_WIN
```

</details>

## Expected Behavior
When pasting text, it should take into account tab indents as well as spaces not only that but it should also do it much faster.

## Current Behavior
When I paste functions and such, I have to wait several seconds/minutes for it to paste and only for it most of the time to not have it's tab indents/spaces taken into accounts. Kind of ironic for a python shell.

## Steps to Reproduce
Copy pasting functions with tab indents or spaces as indents (spaces seem to have a bit less issues?)

**Comment by brokedarius** [comment]
> I can't reproduce this. Please provide your `xonfig` as well as example.

Seems to be only affecting functions with tab indents not spaces after further testing so apologies for lumping them in.

For instance can you please try pasting this small example function in a xonsh shell if you don't mind?: 

```xsh
def save_toml(toml_dict: dict[str, str], toml_path: str) -> None:

	with open(toml_path, 'w') as c:
		c.write(tomlkit.dumps(asdict(toml_dict)))
```

Also I am on windows, tried it on both tabby and windows terminal where its very slow to finish pasting, sometimes it does not when the function is a bit bigger. On conhost it pastes rather fast but the tab indents are not respected either.

**Comment by gforsyth** [comment]
Can you confirm that pasting works as you expect when `xonsh` isn't in the mix?

**Comment by brokedarius** [comment]
yes works fine from regular python console sorry I was away for a little while

**Comment by brokedarius** [comment]
I noticed something:

I opened the project in VSCode and realized that in ptk_shell/key_bindings.py if you change this function from this: 

```xsh
@Condition
def tab_insert_indent():
    """Check if <Tab> should insert indent instead of starting autocompletion.
    Checks if there are only whitespaces before the cursor - if so indent
    should be inserted, otherwise autocompletion.

    """
    before_cursor = get_app().current_buffer.document.current_line_before_cursor

    return bool(before_cursor.isspace())
```

to this:

```xsh
@Condition
def tab_insert_indent():
    """Check if <Tab> should insert indent instead of starting autocompletion.
    Checks if there are only whitespaces before the cursor - if so indent
    should be inserted, otherwise autocompletion.

    """
    before_cursor = get_app().current_buffer.document.current_line_before_cursor

    return True
```

It somehow applies a 'wall patch' to my issue but obviously I lose the ability to tab complete which I don't care for personally but now at least we know that in this function the boolean is most likely always evaluating to False when pasting text because of this expression? `get_app().current_buffer.document.current_line_before_cursor.isspace()`. 

Either that or some interference due to the fact that the TAB key is reserved for other operations? Let me know.

(Only tested on Windows 10)

**Comment by anki-code** [comment]
Hi @brokedarius, I've tested the [example](https://github.com/xonsh/xonsh/issues/5292#issuecomment-1973552607) you provided on Windows 11 with Python 3.13, prompt-toolkit 3.0.52, xonsh 0.22.8:
* Using `cmd.exe`.
* Using new [Windows Terminal](https://github.com/microsoft/terminal).
* Using xonsh code and [PyCharm from dev toolkit](https://github.com/anki-code/xonsh-developer-toolkit?tab=readme-ov-file#pycharm) terminal.

I can't reproduce the issue with paste performance.

There is only one thing that catches my eye - `^I` instead of tab space but after pressing Enter the code succesfully applied.

---------------

I recommend this path to dive into issue on your side:
1. Install [PyCharm from dev toolkit](https://github.com/anki-code/xonsh-developer-toolkit?tab=readme-ov-file#pycharm).
2. Run xonsh in Debug mode.
3. Paste the text and during lag you can (or can't) click Pause button to catch the place in the code.
4. If you can't press the Pause button try to set breakpoint (choose prompt-toolkit source code from PyCharm's Project tree under the "External Libraries" section) and debug step by step:
    * [ptk basic pasting](https://github.com/prompt-toolkit/python-prompt-toolkit/blob/940af53fa443073d9fdca26d5da6cfe6780f6ac9/src/prompt_toolkit/key_binding/bindings/basic.py#L247)
    * [ptk buffer insert_text](https://github.com/prompt-toolkit/python-prompt-toolkit/blob/940af53fa443073d9fdca26d5da6cfe6780f6ac9/src/prompt_toolkit/buffer.py#L1245-L1257)
5. Because of [Windows had issues around speed](https://github.com/xonsh/xonsh/issues/3895) (it's around WSL but still) I suppose that inserting [fires completion](https://github.com/prompt-toolkit/python-prompt-toolkit/blob/940af53fa443073d9fdca26d5da6cfe6780f6ac9/src/prompt_toolkit/buffer.py#L1256) and you're observing slow down.
6. We need to find the concrete place where you observe lag.

I also recommend to try to repeat the issue in `xonsh --no-rc --no-env` and maybe catch the issue in PATH.

**Comment by anki-code** [comment]
I'm going to close this with additional reference to:
* https://github.com/xonsh/xonsh/issues/6259
